enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1090 @@
+	// strategy configuration to be added later.
+}
+
+// ControlPlaneMachineSetStrategyType is an enumeration of different update strategies
+// for the Control Plane Machines.
+type ControlPlaneMachineSetStrategyType string
+
+const (
+	// RollingUpdate is the default update strategy type for a
+	// ControlPlaneMachineSet. This will cause the ControlPlaneMachineSet to
+	// first create a new Machine and wait for this to be Ready
+	// before removing the Machine chosen for replacement.
+	RollingUpdate ControlPlaneMachineSetStrategyType = "RollingUpdate"
+
+	// Recreate causes the ControlPlaneMachineSet controller to first
+	// remove a ControlPlaneMachine before creating its
+	// replacement. This allows for scenarios with limited capacity
+	// such as baremetal environments where additional capacity to
+	// perform rolling updates is not available.
+	Recreate ControlPlaneMachineSetStrategyType = "Recreate"
+
+	// OnDelete causes the ControlPlaneMachineSet to only replace a
+	// Machine once it has been marked for deletion. This strategy
+	// makes the rollout of updated specifications into a manual
+	// process. This allows users to test new configuration on
+	// a single Machine without forcing the rollout of all of their
+	// Control Plane Machines.
+	OnDelete ControlPlaneMachineSetStrategyType = "OnDelete"
+)
+
+// FailureDomain represents the different configurations required to spread Machines
+// across failure domains on different platforms.
+// +union
+type FailureDomains struct {
+	// Platform identifies the platform for which the FailureDomain represents
+	// +unionDiscriminator
+	// +optional
+	Platform configv1.PlatformType `json:"platform,omitempty"`
+
+	// AWS configures failure domain information for the AWS platform
+	// +optional
+	AWS *[]AWSFailureDomain `json:"aws,omitempty"`
+
+	// Azure configures failure domain information for the Azure platform
+	// +optional
+	Azure *[]AzureFailureDomain `json:"azure,omitempty"`
+
+	// GCP configures failure domain information for the GCP platform
+	// +optional
+	GCP *[]GCPFailureDomain `json:"gcp,omitempty"`
+
+	// OpenStack configures failure domain information for the OpenStack platform
+	// +optional
+	OpenStack *[]OpenStackFailureDomain `json:"openstack,omitempty"`
+}
+
+// AWSFailureDomain configures failure domain information for the AWS platform
+type AWSFailureDomain struct {
+	// Subnet is a reference to the subnet to use for this instance
+	// +optional
+	Subnet AWSResourceReference `json:"subnet,omitempty"`
+
+	// Placement configures the placement information for this instance
+	// +optional
+	Placement AWSFailureDomainPlacement `json:"placement,omitempty"`
+}
+
+// AWSFailureDomainPlacement configures the placement information for the AWSFailureDomain
+type AWSFailureDomainPlacement struct {
+	// AvailabilityZone is the availability zone of the instance
+	// +optional
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
```

Comment:
What is the purpose of putting the `availabilityZone` field one level deeper than the similar `zone` fields for other platforms?

Reply:
This is so that it matches the design of the `AWSMachineProviderConfig` from a user perspective, we thought it would be more obvious what the effect of the failure domain is, if it were to look like a subset of the providerSpec

Reply:
Makes sense.